File: 2. Structural Design Patterns/Decorator/Typescript/DecoratorTemplate.ts
// Decorator Pattern – TypeScript Template

abstract class Component {
  abstract operation(): void;
}

class ConcreteComponent extends Component {
  operation(): void {
    /* base work */
  }
}

abstract class BaseDecorator extends Component {
  constructor(protected wrap: Component) {
    super();
  }

  operation(): void {
    this.wrap.operation();
  }
}

class ConcreteDecoratorA extends BaseDecorator {
  operation(): void {
    /* extra A */
    super.operation();
  }
}

class ConcreteDecoratorB extends BaseDecorator {
  operation(): void {
    /* extra B */
    super.operation();
  }
}

(function main() {
  const c: Component = new ConcreteDecoratorB(
    new ConcreteDecoratorA(new ConcreteComponent())
  );
  c.operation();
})();

export {};

File: 2. Structural Design Patterns/Decorator/Go/DecoratorTemplate.go
package main

// Decorator Pattern – Go Template

type Component interface {
    Operation()
}

type ConcreteComponent struct{}
func (ConcreteComponent) Operation() { /* base work */ }

type BaseDecorator struct{ wrap Component }

func (d BaseDecorator) Operation() { d.wrap.Operation() }

type ConcreteDecoratorA struct{ BaseDecorator }

func (c ConcreteDecoratorA) Operation() { /* extra A */ c.BaseDecorator.Operation() }

type ConcreteDecoratorB struct{ BaseDecorator }

func (c ConcreteDecoratorB) Operation() { /* extra B */ c.BaseDecorator.Operation() }

func main() {
    var c Component = ConcreteDecoratorB{
        BaseDecorator{ConcreteDecoratorA{BaseDecorator{ConcreteComponent{}}}},
    }
    c.Operation()
}
